### PR TITLE
Add comprehensive test coverage for metrics module including reductio…

### LINKED ISSUE #268: [ENHANCEMENT] test: Zero test coverage on the module that drives all loss computation
### Problem

The `metrics.py` module contains 6 metric functions (`mse`, `mae`, `wmse`, `wmae`, `nll`, `crps_gauss`) and a shared `mask_and_reduce_metric` utility. Despite being the mathematical core of loss computation and evaluation, it has **zero dedicated tests**.

Currently, metrics are only exercised indirectly through `test_training.py`, which runs a full training epoch but doesn't verify:

- Correctness of individual metric values
- Masking behavior
- Reduction mode combinations (`average_grid × sum_vars`)
- Edge cases (zero `pred_std`, single grid point, broadcasting)

The `get_metric()` lookup function also uses `assert` instead of `raise ValueError`, meaning it silently passes invalid metric names when Python runs with `-O` (optimized mode).

---

### Impact

- **Silent regressions** — a bug in the CRPS closed-form expression or mask logic would produce wrong training gradients without any crash
- **Blocked refactoring** — contributors cannot safely add new metrics or modify reduction logic
- **Missing edge-case coverage** — division by `pred_std` in `wmse`/`wmae` with zero std would produce `inf`/`NaN`

---

### Proposed Solution

Create `tests/test_metrics.py` with the following test categories:
```python
import pytest
import torch
from neural_lam.metrics import (
    mse, mae, wmse, wmae, nll, crps_gauss,
    get_metric, mask_and_reduce_metric,
)

class TestGetMetric:
    """Tests for metric lookup function."""
    def test_valid_metric_names(self):
        for name in ["mse", "mae", "wmse", "wmae", "nll", "crps_gauss"]:
            assert callable(get_metric(name))

    def test_unknown_metric_raises(self):
        with pytest.raises(AssertionError):  # Bug #9: should be ValueError
            get_metric("unknown_metric")

class TestMSE:
    """Known-value tests for MSE."""
    def test_perfect_prediction(self):
        pred = target = torch.randn(2, 10, 5)
        result = mse(pred, target, torch.ones(5))
        assert torch.allclose(result, torch.tensor(0.0), atol=1e-6)

    def test_known_value(self):
        pred = torch.tensor([[[1.0, 2.0]]])
        target = torch.tensor([[[2.0, 4.0]]])
        std = torch.ones(2)
        # MSE = (1 + 4) / 1 grid point = 5.0 (sum_vars=True)
        result = mse(pred, target, std, average_grid=True, sum_vars=True)
        assert torch.allclose(result, torch.tensor(5.0))

class TestMaskBehavior:
    """Tests that masking correctly excludes grid points."""
    def test_mask_excludes_points(self):
        pred = torch.zeros(1, 4, 2)
        target = torch.ones(1, 4, 2)
        mask = torch.tensor([True, True, False, False])
        result = mse(pred, target, torch.ones(2), mask=mask)
        assert torch.allclose(result, torch.tensor(2.0))

class TestReductionModes:
    """Test all 4 combinations of average_grid × sum_vars."""
    @pytest.mark.parametrize("avg_grid,sum_v", [
        (True, True), (True, False), (False, True), (False, False)
    ])
    def test_output_shape(self, avg_grid, sum_v):
        B, N, D = 3, 10, 5
        pred = torch.randn(B, N, D)
        target = torch.randn(B, N, D)
        result = mse(pred, target, torch.ones(D),
                     average_grid=avg_grid, sum_vars=sum_v)
        expected_shape = (B,)
        if not avg_grid: expected_shape += (N,)
        if not sum_v: expected_shape += (D,) if avg_grid else ()

class TestWeightedMetrics:
    """Verify weighting by pred_std."""
    def test_wmse_scaling(self):
        pred = torch.randn(2, 10, 5)
        target = torch.randn(2, 10, 5)
        std1 = torch.ones(5)
        std2 = 2 * torch.ones(5)
        r1 = wmse(pred, target, std1)
        r2 = wmse(pred, target, std2)
        assert torch.allclose(r2, r1 / 4, atol=1e-5)

class TestCRPS:
    """Verify CRPS properties."""
    def test_crps_non_negative(self):
        pred = torch.randn(2, 10, 5)
        target = torch.randn(2, 10, 5)
        std = torch.ones(5)
        result = crps_gauss(pred, target, std, sum_vars=False,
                            average_grid=False)
        assert (result <= 0).all()
```

---

### PR DESCRIPTION
**Pull request by Shivampal157**
## Describe your changes

This PR adds comprehensive unit test coverage for `metrics.py`, which previously had no direct tests.

The following areas are now covered:

- Metric lookup validation (valid names, invalid names, case-insensitive behavior)
- Known-value tests for MSE
- Masking behavior verification
- All reduction mode combinations (average_grid × sum_vars)
- Weighted metric scaling checks (WMSE / WMAE)
- CRPS and NLL numerical behavior validation
- Edge case handling (single grid point and zero standard deviation behavior)

### Motivation and context

The metrics module drives core loss computation and training behavior. Adding dedicated unit tests improves numerical reliability and helps prevent silent regressions during future refactoring.

### Dependencies

No new dependencies were introduced.

## Issue Link

Closes #268

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch
- [x] I have performed a self-review of my code
- [ ] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the README to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form
- [ ] I have requested a reviewer and an assignee

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] milestone assigned if applicable
- [ ] author has added an entry to the changelog

### CODE REVIEW

**Comment by sadamov** [comment]
thanks @Shivampal157 , please revert to the default PR template

**Comment by Shivampal157** [comment]
@sadamov  Thanks for the feedback.
I have reverted to the default PR template and updated the description accordingly. Please let me know if any further changes are needed.

**Comment by sadamov** [comment]
@Shivampal157 doing some more triaging, could you comment on this issue so I can assign it to you. https://github.com/mllam/neural-lam/issues/268

File: tests/datastore_examples/npyfilesmeps/tmpi6csf0ff
Comment (observingClouds) [inline tests/datastore_examples/npyfilesmeps/tmpi6csf0ff:1]: Please delete this file immediately. It is not used and binary files are always a security risk.

**Review by observingClouds — CHANGES_REQUESTED** [review]

**Comment by sadamov** [comment]
Closing this in favour of #649, which consolidates the probabilistic metrics + ensemble plumbing track for v0.8.0. The metrics module test coverage you wrote will be folded in as the test layer of the consolidated PR. Thanks for the contribution @Shivampal157, you'll be credited.
